lib/ts/Scrubber.cc

```diff
@@ -0,0 +1,201 @@
+    // Error("Scrubbing pattern failed at offset %d: %s.", erroroffset, error);
+    return false;
+  }
+
+  // add the scrub pattern to our list
+  s              = new Scrub;
+  s->pattern     = pattern;
+  s->replacement = replacement;
+  s->compiled_re = re;
+  scrubs.push_back(s);
+
+  return true;
+}
+
+void
+Scrubber::scrub_buffer(char *buffer) const
+{
+  // apply every Scrub in the vector, in order
+  for (auto s : scrubs) {
+    scrub_buffer(buffer, s);
+  }
+}
+
+void
+Scrubber::scrub_buffer(char *buffer, Scrub *scrub) const
+{
+  char *buffer_ptr;
+  int num_matched;
+  int match_len;
+  int replacement_len;
+  int buffer_len = strlen(buffer);
+
+  // execute regex
+  num_matched = pcre_exec(scrub->compiled_re, nullptr, buffer, buffer_len, 0, 0, scrub->ovector, scrub->OVECCOUNT);
+  if (num_matched < 0) {
+    switch (num_matched) {
+    case PCRE_ERROR_NOMATCH:
+      return;
+    default:
+      // Error("PCRE matching error %d\n", num_matched);
+      break;
+    }
+  }
+
+  /*
+   * When scrubbing the buffer in place, there are 2 scenarios we need to consider:
+   *
+   *   1) The replacement text length is shorter or equal to the text we want to scrub away
+   *   2) The replacement text is longer than the text we want to scrub away
+   *
+   * In case 1, we simply "slide" everything left a bit. Our final buffer should
+   * look like this (where XXXX is the replacment text):
+   *
+   *                                new_end  orig_end
+   *                                    V      V
+   *   -----------------------------------------
+   *   |ORIGINAL TEXT|XXXX|ORIGINAL TEXT|      |
+   *   -----------------------------------------
+   *
+   * In case 2, since the final buffer would be longer than the original allocated buffer,
+   * we need to truncate everything that would have run over the original end of the buffer.
+   * The final buffer should look like this:
+   *
+   *                                         new_end
+   *                                        orig_end
+   *                                           V
+   *   -----------------------------------------
+   *   |ORIGINAL TEXT|XXXXXXXXXXXXXXXXXXX|ORIGI|NAL TEXT
+   *   -----------------------------------------
+   *
+   */
+
+  buffer_ptr      = buffer;
+  match_len       = scrub->ovector[1] - scrub->ovector[0];
+  replacement_len = scrub->replacement.size();
+
+  if (replacement_len <= match_len) { // case 1
+    // overwrite the matched text with the replacement text
+    buffer_ptr += scrub->ovector[0];
+    memcpy(buffer_ptr, scrub->replacement.ptr(), replacement_len);
+    buffer_ptr += replacement_len;
+
+    // slide everything after the matched text left to fill the gap including the '\0'
+    memmove(buffer_ptr, buffer + scrub->ovector[1], buffer_len - scrub->ovector[1] + 1);
+
+    // the last char should ALWAYS be a '\0' otherwise we did our math wrong
+    ink_assert(buffer[strlen(buffer)] == '\0');
```

Comment:
This is expensive and pointless - by definition, `strlen` will find a null so this will always be true, unless it segfaults. The index should be computed from the other lengths, not searched for, if you want to verify your math.